src/x11.cc

```diff
@@ -307,6 +310,57 @@ __attribute__((noreturn)) static int x11_ioerror_handler(Display *d) {
   CRIT_ERR("X IO Error: Display %lx\n", reinterpret_cast<uint64_t>(d));
 }
 
+/// @brief Function to get virtual root windows of screen.
+///
+/// Some WMs (swm, tvtwm, amiwm, enlightenment, etc.) use virtual roots to
+/// manage workspaces. These are direct descendants of root and WMs reparent all
+/// children to them.
+///
+/// @param screen screen to get the (current) virtual root of @return the
+/// virtual root window of the screen
+static Window VirtualRootWindowOfScreen(Screen *screen) {
+  Window root = screen->root;
+  Display *dpy = screen->display;
+
+  Window rootReturn, parentReturn, *children;
+  unsigned int numChildren;
+  Atom actual_type;
+  int actual_format;
+  unsigned long nitems, bytesafter;
+
+  /* go look for a virtual root */
+  Atom __SWM_VROOT = ATOM(__SWM_VROOT);
+  if (XQueryTree(dpy, root, &rootReturn, &parentReturn, &children,
+                 &numChildren)) {
+    for (int i = 0; i < numChildren; i++) {
+      Window *newRoot = None;
+
+      if (XGetWindowProperty(
+              dpy, children[i], __SWM_VROOT, 0, 1, False, XA_WINDOW,
+              &actual_type, &actual_format, &nitems, &bytesafter,
+              reinterpret_cast<unsigned char **>(&newRoot)) == Success &&
+          newRoot != None) {
+        root = *newRoot;
+        break;
+      }
+    }
+    if (children) XFree((char *)children);
+  }
+
+  return root;
+}
+
+// Polyfill overwrite default Xlib definitions to account for virtual root
+// windows
+#undef RootWindowOfScreen
+#define RootWindowOfScreen(s) VirtualRootWindowOfScreen(s)
+#undef RootWindow
+#define RootWindow(dpy, screen) \
+  VirtualRootWindowOfScreen(ScreenOfDisplay(dpy, screen))
+#undef DefaultRootWindow
+#define DefaultRootWindow(dpy) \
+  VirtualRootWindowOfScreen(DefaultScreenOfDisplay(dpy))
```

Comment:
Moved handling of virtual roots from `find_desktop_window` here and overwrote Xlib macros to use it for consistency.

Overwriting macros means that reading `wm-spec` atoms requires accessing (real) root though `screen->root`, so I'm not 100% sure about doing that.